src/layers/CustomDisplayObjects/ComplexRopeGeometry.ts

```diff
@@ -0,0 +1,212 @@
+
+      for (let j = 0; j < total; j++) {
+        // time to do some smart drawing!
+
+        // calculate pixel distance from previous point
+        const dx = prev.x - segments[i].points[j].x;
+        const dy = prev.y - segments[i].points[j].y;
+        const distance = Math.sqrt(dx * dx + dy * dy);
+
+        prev = segments[i].points[j];
+        amount += distance / textureWidth;
+
+        uvs[uvIndex] = amount;
+        uvs[uvIndex + 1] = 0.5 - radius;
+
+        uvs[uvIndex + 2] = amount;
+        uvs[uvIndex + 3] = 0.5 + radius;
+        uvIndex += 4;
+      }
+
+      for (let j = 0; j < total - 1; j++) {
+        indices[indexCount++] = indicesIndex;
+        indices[indexCount++] = indicesIndex + 1;
+        indices[indexCount++] = indicesIndex + 2;
+
+        indices[indexCount++] = indicesIndex + 2;
+        indices[indexCount++] = indicesIndex + 1;
+        indices[indexCount++] = indicesIndex + 3;
+        indicesIndex += 2;
+      }
+      indicesIndex += 2;
+    }
+
+    // ensure that the changes are uploaded
+    uvBuffer.update();
+    indexBuffer.update();
+
+    this.updateVertices();
+  }
+
+  /** refreshes vertices of Rope mesh */
+  public updateVertices(): void {
+    const segments = this.segments;
+    const pointCount = sum(segments, (segment) => segment.points.length);
+
+    if (pointCount < 1) {
+      return;
+    }
+
+    const segmentCount = segments.length;
+    let lastIndex = 0;
+    for (let i = 0; i < segmentCount; i++) {
+      let lastPoint = segments[i].points[0];
+      let nextPoint;
+      let perpX = 0;
+      let perpY = 0;
+
+      const vertices = this.buffers[0].data;
+      const total = segments[i].points.length;
+      let index = 0;
+      for (let j = 0; j < total; j++) {
+        const point = segments[i].points[j];
+        index = lastIndex + j * 4;
+
+        if (j < segments[i].points.length - 1) {
+          nextPoint = segments[i].points[j + 1];
+        } else {
+          nextPoint = point;
+        }
+
+        perpY = -(nextPoint.x - lastPoint.x);
+        perpX = nextPoint.y - lastPoint.y;
+
+        const perpLength = Math.sqrt(perpX * perpX + perpY * perpY);
+        const num = segments[i].diameter / 2;
+
+        perpX /= perpLength;
+        perpY /= perpLength;
+
+        perpX *= num;
+        perpY *= num;
+
+        vertices[index] = point.x + perpX;
+        vertices[index + 1] = point.y + perpY;
+        vertices[index + 2] = point.x - perpX;
+        vertices[index + 3] = point.y - perpY;
+        lastPoint = point;
+      }
+      lastIndex = index + 4;
+    }
+
+    this.buffers[0].update();
+  }
+
+  public update(): void {
+    // TODO check what we need here
+    // this.build() calls this.updateVertices()
+    this.build();
+
+    // if (this.textureScale > 0) {
+    // this.build(); // we need to update UVs
+    // } else {
+    //   this.updateVertices();
+    // }
+
+    // this.updateVertices();
+  }
+}
```

Comment:
Is this still a TODO as well?

Reply:
I have improved the comment but kept the todo. Depending on the performance we end up with after all changes we have planned I want to know where to look for possible quick performance gains if needed.

There are a couple potential optimizations to be found around when and how much to update on every render
For instance; what needs to be recalculated on zoom/pan? maybe nothing and the build step can be skipped if nothing has changed in the segments or texture scale
How about z-factor? Might need to force rebuild everything on z-factor change, or maybe everything is recalculated from CasingAndCementLayer anyway in that instance.